File: src/test/java/org/cerion/stocklist/functions/conditions/PriceConditionTest.java
package org.cerion.stocklist.functions.conditions;

import org.cerion.stocklist.arrays.BandArray;
import org.cerion.stocklist.functions.FunctionTestBase;
import org.cerion.stocklist.functions.IPriceOverlay;
import org.cerion.stocklist.functions.conditions.Condition;
import org.cerion.stocklist.functions.conditions.PriceCondition;
import org.cerion.stocklist.overlays.BollingerBands;
import org.cerion.stocklist.overlays.SimpleMovingAverage;
import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;


public class PriceConditionTest extends FunctionTestBase {

    @Test(expected = IllegalArgumentException.class)
    public void verifiesCondition() {
        new PriceCondition(Condition.INSIDE, new SimpleMovingAverage());
    }

    @Test
    public void floatOverlay() {
        testCondition(Condition.BELOW, new SimpleMovingAverage(20));
        testCondition(Condition.ABOVE, new SimpleMovingAverage(94));
    }

    @Test
    public void bandOverlay() {
        testCondition(Condition.INSIDE, new BollingerBands(20, 2));
        testCondition(Condition.ABOVE, new BollingerBands(94, 0.2));
        testCondition(Condition.BELOW, new BollingerBands(30, 0.2));
    }

    @Test
    public void toStringTest() {
        assertEquals("Price above SMA 29", new PriceCondition(Condition.ABOVE, new SimpleMovingAverage(29)).toString() );
        assertEquals("Price below SMA 29", new PriceCondition(Condition.BELOW, new SimpleMovingAverage(29)).toString() );
        assertEquals("Price inside BB 30,3.0", new PriceCondition(Condition.INSIDE, new BollingerBands(30, 3.0)).toString() );
    }

    private void testCondition(Condition trueCondition, IPriceOverlay overlay) {
        for(Condition c : Condition.values()) {
            if (c == Condition.INSIDE && overlay.getResultType() != BandArray.class)
                continue;

            PriceCondition condition = new PriceCondition(c, overlay);
            if (c == trueCondition)
                assertTrue("Price " + trueCondition + " with " + overlay, condition.eval(mPriceList));
            else
                assertFalse("Price NOT " + trueCondition + " with " + overlay, condition.eval(mPriceList));
        }
    }
}

File: src/test/kotlin/functions/conditions/PriceConditionTest.kt
package org.cerion.stocks.core.functions.conditions

import org.cerion.stocks.core.TestBase
import org.cerion.stocks.core.arrays.BandArray
import org.cerion.stocks.core.functions.IPriceOverlay
import org.cerion.stocks.core.overlays.BollingerBands
import org.cerion.stocks.core.overlays.SimpleMovingAverage
import org.junit.Assert.*
import org.junit.Test


class PriceConditionTest : TestBase() {

    @Test(expected = IllegalArgumentException::class)
    fun verifiesCondition() {
        PriceCondition(Condition.INSIDE, SimpleMovingAverage())
    }

    @Test
    fun floatOverlay() {
        testCondition(Condition.BELOW, SimpleMovingAverage(20))
        testCondition(Condition.ABOVE, SimpleMovingAverage(94))
    }

    @Test
    fun bandOverlay() {
        testCondition(Condition.INSIDE, BollingerBands(20, 2.0))
        testCondition(Condition.ABOVE, BollingerBands(94, 0.2))
        testCondition(Condition.BELOW, BollingerBands(30, 0.2))
    }

    @Test
    fun toStringTest() {
        assertEquals("Price above SMA 29", PriceCondition(Condition.ABOVE, SimpleMovingAverage(29)).toString())
        assertEquals("Price below SMA 29", PriceCondition(Condition.BELOW, SimpleMovingAverage(29)).toString())
        assertEquals("Price inside BB 30,3.0", PriceCondition(Condition.INSIDE, BollingerBands(30, 3.0)).toString())
    }

    private fun testCondition(trueCondition: Condition, overlay: IPriceOverlay) {
        for (c in Condition.values()) {
            if (c == Condition.INSIDE && overlay.resultType != BandArray::class.java)
                continue

            val condition = PriceCondition(c, overlay)
            if (c == trueCondition)
                assertTrue("Price $trueCondition with $overlay", condition.eval(priceList))
            else
                assertFalse("Price NOT $trueCondition with $overlay", condition.eval(priceList))
        }
    }
}